src/zjs_callbacks.c

```diff
@@ -248,33 +235,33 @@ int32_t zjs_add_callback_list(jerry_value_t js_func,
     }
 }
 
-int32_t add_callback(jerry_value_t js_func,
-                     jerry_value_t this,
-                     void* handle,
-                     zjs_post_callback_func post,
-                     uint8_t once)
+zjs_callback_id add_callback(jerry_value_t js_func,
+                             jerry_value_t this,
+                             void* handle,
+                             zjs_post_callback_func post,
+                             uint8_t once)
 {
-    struct zjs_callback_map* new_cb = zjs_malloc(sizeof(struct zjs_callback_map));
+    struct zjs_callback_t* new_cb = zjs_malloc(sizeof(struct zjs_callback_t));
     if (!new_cb) {
         DBG_PRINT("error allocating space for new callback\n");
         return -1;
     }
-    new_cb->js = zjs_malloc(sizeof(struct zjs_callback_t));
-    if (!new_cb->js) {
-        DBG_PRINT("error allocating space for new callback\n");
-        zjs_free(new_cb);
-        return -1;
-    }
-    new_cb->type = CALLBACK_TYPE_JS;
+    memset(new_cb, 0, sizeof(struct zjs_callback_t));
+
+    new_cb->flags |= (CALLBACK_TYPE_JS << TYPE_BIT);
+    new_cb->flags |= ((once) ? 1 : 0 << ONCE_BIT);
     new_cb->id = new_id();
-    new_cb->js->js_func = jerry_acquire_value(js_func);
-    new_cb->js->this = jerry_acquire_value(this);
-    new_cb->js->post = post;
-    new_cb->js->handle = handle;
-    new_cb->js->func_list = NULL;
-    new_cb->js->max_funcs = 0;
-    new_cb->js->num_funcs = 0;
-    new_cb->js->once = once;
+    // For a single callback, instead of allocating for just one function, we
+    // can use the pointer to hold the function since it will never grow
+    jerry_value_t func_val = jerry_acquire_value(js_func);
+    memcpy(&new_cb->func_list, &func_val, sizeof(jerry_value_t));
```

Comment:
oh, ya I hadn't thought of that. Didn't like my nasty hack? :) ill add its own value to the union, much better way of doing it